src/core/util/kdtree.h

```diff
@@ -0,0 +1,248 @@
+};
+
+template <SubscriptableOrder PointType>
+class KDTree;
+
+// @brief k-dimensional Rectangle used for query search
+template <SubscriptableOrder PointType>
+struct Rect {
+    enum bound_type { kOpen = 0, kClosed };
+
+    PointType lower_bound_, upper_bound_;
+    std::vector<bound_type> lower_bound_type_, upper_bound_type_;
+
+    Rect() = default;
+    Rect(Rect<PointType> const&) = default;
+    Rect(Rect<PointType>&&) = default;
+    Rect<PointType>& operator=(Rect<PointType> const&) = default;
+
+    Rect(PointType lower_bound, PointType upper_bound)
+        : lower_bound_(std::move(lower_bound)), upper_bound_(std::move(upper_bound)) {
+        if (lower_bound.GetDim() != upper_bound.GetDim())
+            throw std::logic_error("Dimensionalities of given points don't match");
+
+        lower_bound_type_ = std::vector<bound_type>(lower_bound.GetDim(), kOpen);
+        upper_bound_type_ = std::vector<bound_type>(upper_bound.GetDim(), kOpen);
+    }
+
+    void SetBoundType(std::vector<bound_type> lower_bound_type,
+                      std::vector<bound_type> upper_bound_type) {
+        if (lower_bound_.GetDim() != upper_bound_.GetDim())
+            throw std::logic_error("Dimensionalities of given points don't match");
+
+        lower_bound_type_ = std::move(lower_bound_type);
+        upper_bound_type_ = std::move(upper_bound_type);
+    }
+
+    bool Fits(PointType const& point) const {
+        if (lower_bound_.GetDim() != upper_bound_.GetDim())
+            throw std::logic_error("Dimensionalities of a given point and boundaries don't match");
+
+        for (size_t i = 0; i < point.GetDim(); i++) {
+            if (lower_bound_type_[i] == kOpen) {
+                if (point[i] <= lower_bound_[i]) return false;
+            } else if (point[i] < lower_bound_[i]) {
+                return false;
+            }
+
+            if (upper_bound_type_[i] == kOpen) {
+                if (point[i] >= upper_bound_[i]) return false;
+            } else if (point[i] > upper_bound_[i]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+};
+
+template <SubscriptableOrder PointType>
+class KDTree {
+private:
+    class Node {
+        friend class KDTree;
+
+        PointType point_;
+        std::unique_ptr<Node> left_;
+        std::unique_ptr<Node> right_;
+        size_t axis_;
+
+    public:
+        Node() = default;
+
+        template <typename U>
+        Node(U&& point, std::unique_ptr<Node> left, std::unique_ptr<Node> right, size_t axis)
+            : point_(std::move(point)),
+              left_(std::move(left)),
+              right_(std::move(right)),
+              axis_(axis) {}
+
+        ~Node() = default;
+
+        Node(Node&&) = delete;
+        Node(Node const&) = delete;
+        Node& operator=(Node const&) = delete;
+        Node& operator=(Node&&) = delete;
+    };
+
+    std::unique_ptr<Node> root_;
+    size_t size_;
+
+    void Erase(std::unique_ptr<Node>& dest);
```

Comment:
`Erase` is only used in the destructor and the default destructor does the same thing. I think it should be removed.